quarkchain/cluster/miner.py

```diff
@@ -251,42 +261,39 @@ def _mine_new_block_async(self):
             return None
         return asyncio.ensure_future(mine_new_block())
 
-    async def get_work(self, now=None) -> (MiningWork, Block):
+    async def get_work(
+        self, now=None, coinbase_addr: Optional[Address] = None
+    ) -> (MiningWork, Block):
         if not self.remote:
             raise ValueError("Should only be used for remote miner")
 
         if now is None:  # clock open for mock
             now = time.time()
-        work_prev_hash = (
-            self.current_work.header.hash_prev_block if self.current_work else None
-        )
+
+        block = None
+        header_hash = self.current_works.get(coinbase_addr)
+        if header_hash:
+            block = self.work_map.get(header_hash)
         tip_hash = self.get_header_tip_func().get_hash()
         if (
-            not self.current_work  # no cache
-            or work_prev_hash != tip_hash  # cache outdated
-            or now - self.current_work.header.create_time > 5  # stale
+            not block  # no work cache
+            or block.header.hash_prev_block != tip_hash  # cache outdated
+            or now - block.header.create_time > 10  # stale
         ):
-            block = await self.create_block_async_func(retry=False)
+            block = await self.create_block_async_func(
+                retry=False, coinbase_addr=coinbase_addr
+            )
             if not block:
                 raise RuntimeError("Failed to create block")
-            self.current_work = block
-
-        header = self.current_work.header
-        header_hash = header.get_hash_for_mining()
-        # store in memory for future retrieval during work submission
-        self.work_map[header_hash] = self.current_work
-
-        # clean up worker map
-        # TODO: for now, same param as go-ethereum. or LRU cache?
-        self.work_map = {
-            h: b
-            for h, b in self.work_map.items()
-            if now - b.header.create_time < 7 * 12
-        }
+            header_hash = block.header.get_hash_for_mining()
+            # note that key could be None
+            self.current_works[coinbase_addr] = header_hash
```

Comment:
If coinbase_addr is None, why not use configed coinbase?  Using None as the key is very confusing.

Reply:
because otherwise we have to pass the default coinbase (the one from the config) to `class Miner`, which I think is unnecessary